plutus-core/untyped-plutus-core/src/Language/UntypedPlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -288,18 +284,23 @@ instance MonadEmitter (CekCarryingM term uni fun s) where
 -- @Ix fun@ which implies @Ord fun@ which implies @Eq fun@.
 instance (Eq fun, Hashable fun, ToExMemory term) =>
             SpendBudget (CekCarryingM term uni fun s) fun (ExBudgetCategory fun) term where
-    spendBudget key budget = do
-        modifying exBudgetStateTally
-                (<> (ExTally (singleton key budget)))
-        newBudget <- exBudgetStateBudget <%= (<> budget)
-        mode <- asks cekEnvBudgetMode
-        case mode of
-            Counting -> pure ()
-            Restricting resb ->
-                when (exceedsBudget resb newBudget) $
-                    throwingWithCause _EvaluationError
-                        (UserEvaluationError $ CekOutOfExError resb newBudget)
-                        Nothing  -- No value available for error
+    spendBudget key budgetSpent = do
+        exSt <- get
+        case exSt of
+            CountingSt budgetTotal -> do
+                let budgetTotal' = budgetTotal <> budgetSpent
+                put $! CountingSt budgetTotal'
+            TallyingSt tallies budgetTotal -> do
+                let budgetTotal' = budgetTotal <> budgetSpent
+                put $! TallyingSt (tallies <> ExTally (singleton key budgetSpent)) budgetTotal'
+            RestrictingSt budgetLeft -> do
+                let budgetLeft' = budgetLeft `minusExBudget` budgetSpent
+                if exceedsBudget budgetLeft budgetSpent
```

Comment:
Alternatively, we could check whether that `budgetLeft' < 0`? That avoids confusingly using `budgetLeft` after defining `budgetLeft'`

Reply:
> Alternatively, we could check whether that `budgetLeft' < 0`?

Done (replaced `exceedsBudget` with `isNegativeBudget`).